File: src/client/python/src/transfer.py
from src.network.network import get, post
from sessionless import SessionlessSecp256k1
import time

async def transfer(uuid, destinationUUID, nineumUniqueIds, price, currency, get_keys):
    sessionless = SessionlessSecp256k1()

    timestamp = f'{round(time.time() * 1000)}'
    message = f'{timestamp}{uuid}{destinationUUID}{''.join(nineumUniqueIds)}{price}{currency}'
    signature = await sessionless.sign(message.encode('ascii'), get_keys)

    payload = {
        "timestamp": timestamp,
        "uuid": uuid,
        "destinationUUID": destinationUUID,
        "nineumUniqueIds": nineumUniqueIds,
        "price": price,
        "currency": currency,
        "signature": signature
    }

    response = await post(f'user/{uuid}/transfer', payload)

    return response

File: src/server/node/src/routes/transfer.js
import user from '../user/user.js';
import nineum from '../nineum/nineum.js';
import sessionless from 'sessionless-node';

const transfer = async (req, res) => {
  const body = req.body;
  const timestamp = body.timestamp;
  const uuid = req.params.uuid;
  const destinationUUID = body.destinationUUID;
  const nineumUniqueIds = body.nineumUniqueIds;
  const price = body.price;
  const currency = body.currency;
  const signature = body.signature;

  const message = timestamp + uuid + destinationUUID + nineumUniqueIds.join('') + price + currency;
console.log(message);
  const sourceUser = await user.getUser(uuid);
console.log('got source user');

  if(!sessionless.verifySignature(signature, message, sourceUser.pubKey)) {
console.log('auth failed');
    res.status(403);
    return res.send({error: 'Auth error'});
  }

  if(price && price !== 0) {
console.log('unimplemented');
    res.status(501);
    return res.send({error: 'unimplemented'});
  }

  const destinationUser = await user.getUser(destinationUUID);

  const updatedUser = await nineum.transferNineum(sourceUser, destinationUser, nineumUniqueIds, price, currency);

console.log('sending updatedUser', updatedUser);

  res.send(updatedUser);
};

export { transfer };
